### PR TITLE
Pls bench

### PR DESCRIPTION
**Pull request by oojo12**
This adds PLS benchmarking code. Albeit, I made a mistake and made it off the iai_ols_bench branch so it is dependent on a merge of that branches changes to the make_dataset function. The criterion diagrams are attached as [linfa_pls.zip](https://github.com/rust-ml/linfa/files/9946094/linfa_pls.zip)

### Context
1. Laptop idle except bench runs
2. Laptop mostly full charge but running on battery (not plugged in)
3. Rust Version - rustc 1.65.0 (897e37553 2022-11-02)

### Bench Results cargo bench -p linfa-pls
```
     Running benches\pls.rs (target\release\deps\pls-e5f47fa7aee48589.exe)
Linfa_pls/SVD/1000      time:   [519.14 µs 535.71 µs 553.07 µs]
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe
Linfa_pls/SVD/10000     time:   [5.1048 ms 5.1683 ms 5.2393 ms]
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe
Benchmarking Linfa_pls/SVD/100000: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.3s, or reduce sample count to 70.
Linfa_pls/SVD/100000    time:   [60.800 ms 62.163 ms 63.621 ms]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
```

### Bench Results cargo bench -p linfa-pls -F intel-mkl-static
```
     Running benches\pls.rs (target\release\deps\pls-171ffb7df7822c7b.exe)
Linfa_pls/SVD/1000      time:   [524.84 µs 537.65 µs 552.99 µs]
                        change: [+8.4643% +12.173% +15.848%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
Linfa_pls/SVD/10000     time:   [4.9096 ms 4.9755 ms 5.0472 ms]
                        change: [-5.4899% -3.7303% -1.8338%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
Benchmarking Linfa_pls/SVD/100000: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.2s, or reduce sample count to 80.
Linfa_pls/SVD/100000    time:   [58.798 ms 59.235 ms 59.703 ms]
                        change: [-7.0392% -4.7092% -2.4666%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild
```

### CODE REVIEW

File: algorithms/linfa-pls/Cargo.toml [hunk 1]
```diff
@@ -39,3 +39,9 @@ linfa = { version = "0.6.0", path = "../.." }
 linfa-datasets = { version = "0.6.0", path = "../../datasets", features = ["linnerud"] }
 approx = "0.4"
 rand_xoshiro = "0.6"
+criterion = "0.4.0"
+statrs = "0.16.0"
+
+[[bench]]
+name = "pls"
+harness = false
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/Cargo.toml:47]: add newline

**Comment by oojo12** [comment]
I want to do the below to bench both `SVD` and `Nipal` algorithm. 

```
use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use linfa::traits::Fit;
use linfa::Dataset;
use linfa_datasets::generate::make_dataset;
use linfa_pls::pls_generic::Algorithm;
use linfa_pls::PlsRegression;
use statrs::distribution::{DiscreteUniform, Laplace};

#[allow(unused_must_use)]
fn perform_pls(num_rows: usize, alg: Algorithm) {
    let feat_distr = Laplace::new(0.5, 5.).unwrap();
    let target_distr = DiscreteUniform::new(0, 5).unwrap();
    let num_feats: usize = 5;
    let dataset = make_dataset(num_rows, num_feats, feat_distr, target_distr);
    let targets = dataset.targets.into_shape((num_rows, 1)).unwrap();
    let features = dataset.records;
    let dataset = Dataset::new(features, targets);
    let model = PlsRegression::params(3)
        .scale(true)
        .max_iterations(200)
        .algorithm(alg);
    model.fit(&dataset);
}

fn bench(c: &mut Criterion) {
    for (alg, name) in [(Algorithm::Nipals, "Nipals"), (Algorithm::Svd, "Svd")] {
        let mut group = c.benchmark_group("Linfa_pls");
        let sizes: [usize; 3] = [1_000, 10_000, 100_000];
        for size in sizes {
            let input = (size, alg);
            group.bench_with_input(BenchmarkId::new(name, size), &input, |b, (size, alg)| {
                b.iter(|| perform_pls(*size, *alg));
            });
        }
        group.finish();
    }
}

criterion_group!(benches, bench);
criterion_main!(benches);
```

However, this gives an error because `pls_generic` is a private module can I make it public?

File: algorithms/linfa-pls/benches/pls.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#[allow(unused_must_use)]
+fn perform_pls(num_rows: usize) {
+    let feat_distr = Laplace::new(0.5, 5.).unwrap();
+    let target_distr = DiscreteUniform::new(0, 5).unwrap();
+    let num_feats: usize = 5;
+    let dataset = make_dataset(num_rows, num_feats, feat_distr, target_distr);
+    let targets = dataset.targets.into_shape((num_rows, 1)).unwrap();
+    let features = dataset.records;
+    let dataset = Dataset::new(features, targets);
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/benches/pls.rs:16]: Move the dataset generation outside of the method

File: algorithms/linfa-pls/benches/pls.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    let model = PlsRegression::params(3).scale(true).max_iterations(200);
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/benches/pls.rs:17]: Can you also run with `PlsCca` and `PlsCanonical`?

File: algorithms/linfa-pls/benches/pls.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+use linfa_pls::PlsRegression;
+use statrs::distribution::{DiscreteUniform, Laplace};
+
+#[allow(unused_must_use)]
+fn perform_pls(num_rows: usize) {
+    let feat_distr = Laplace::new(0.5, 5.).unwrap();
+    let target_distr = DiscreteUniform::new(0, 5).unwrap();
+    let num_feats: usize = 5;
+    let dataset = make_dataset(num_rows, num_feats, feat_distr, target_distr);
+    let targets = dataset.targets.into_shape((num_rows, 1)).unwrap();
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/benches/pls.rs:14]: Maybe we should include multi-target datasets in the benchmarks.

**Comment by YuhanLiin** [comment]
For `Algorithm` you can do add `pub use pls_generic::*`

File: algorithms/linfa-pls/benches/pls.rs (hunk 4 above)
Comment (oojo12) [reply algorithms/linfa-pls/benches/pls.rs:14]: I can see if it is a seamless addition

File: algorithms/linfa-pls/benches/pls.rs (hunk 3 above)
Comment (oojo12) [reply algorithms/linfa-pls/benches/pls.rs:17]: sure thing.

File: algorithms/linfa-pls/benches/pls.rs (hunk 2 above)
Comment (oojo12) [reply algorithms/linfa-pls/benches/pls.rs:16]: will do.

File: algorithms/linfa-pls/benches/pls.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        let feat_distr = Laplace::new(0.5, 5.).unwrap();
+        let target_distr = DiscreteUniform::new(0, 5).unwrap();
+
+        for size in sizes {
+            let dataset = make_dataset(size, 5, feat_distr, target_distr);
+            let targets = dataset.targets.into_shape((size, 1)).unwrap();
+            let features = dataset.records;
+            let dataset = Dataset::new(features, targets);
+            let input = (dataset, alg);
+            group.bench_with_input(BenchmarkId::new(name, size), &input, |b, (dataset, alg)| {
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/benches/pls.rs:32]: Mention the algorithm in the benchmark ID

File: algorithms/linfa-pls/benches/pls.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+        let mut pls_regression_id = "Regression-".to_string();
+        pls_regression_id.push_str(name);
+        let mut pls_canonical_id = "Canonical-".to_string();
+        pls_canonical_id.push_str(name);
+        let mut pls_cca_id = "Cca-".to_string();
+        pls_cca_id.push_str(name);
+
+        for size in sizes {
+            for num_feat in num_feats {
```
Comment (YuhanLiin) [inline algorithms/linfa-pls/benches/pls.rs:52]: Now that I think about it, we don't need to permute all sample sizes and feature counts. We only really need (1000, 5), (10k, 5), (100k, 5), and (100k, 10). In other words, high feature counts are only really interesting when sample count is high. That will reduce the total number of benchmarks by 12, which is alot.

Comment (oojo12) [reply algorithms/linfa-pls/benches/pls.rs]: added it here.

**Review by YuhanLiin — APPROVED** [review]

**Pull request merged**